lib/Conversion/TritonGPUToLLVM/Utility.cpp

```diff
@@ -230,14 +230,36 @@ static Value commonShflSync(Location loc, ConversionPatternRewriter &rewriter,
 #endif
 }
 
+static Value shflUp_amd(Location loc, ConversionPatternRewriter &rewriter,
+                        Value val, int i, Value laneId) {
+  GCNBuilder builder;
+  Value mask = icmp_slt(laneId, i32_val(i));
+  Value delta = sub(laneId, i32_val(i));
+  Value index = select(mask, laneId, delta);
+  Value byteOffset = i32_val(2);
+  Value permuteAddr = shl(index, byteOffset);
+  auto shfl = builder.create("ds_bpermute_b32");
+  auto dOpr = builder.newOperand("=v");
+  auto addrOpr = builder.newOperand(permuteAddr, "v");
+  auto aOpr = builder.newOperand(val, "v");
+  (*shfl)(dOpr, addrOpr, aOpr);
+  auto swait = builder.create("s_waitcnt lgkmcnt(0)");
+  (*swait)();
+  return builder.launch(rewriter, loc, val.getType(), true);
+}
+
 Value shflSync(Location loc, ConversionPatternRewriter &rewriter, Value val,
                int i) {
   return commonShflSync(loc, rewriter, val, i, "bfly", "0x1f");
 }
 
 Value shflUpSync(Location loc, ConversionPatternRewriter &rewriter, Value val,
-                 int i) {
+                 int i, Value laneId) {
+#ifdef USE_ROCM
+    return shflUp_amd(loc, rewriter, val, i, laneId);
+#else
   return commonShflSync(loc, rewriter, val, i, "up", "0x0");
+#endif
```

Comment:
What do you think about moving this switch in `commonShflSync` function?

My argument is following:

Currently `commonShflSync` function support two modes: "up" and "bfly", but our current implementation support only "bfly" (and do not report any errors in case or "up" :slightly_frowning_face:  )

To preserve same upstream semantics we should support "up" in `commonShflSync` and insert an assert for new  shuffle types. This way we could save some debugging time in case upstream uses `commonShflSync` in new places.

Reply:
Yeah. I was debating where to introduce the shfl_up code for AMD. Your suggestion makes sense. I'll change it later.

Reply:
I agree this is a good idea.